django/contrib/contenttypes/models.py

```diff
@@ -174,20 +174,23 @@ def model_class(self):
         except LookupError:
             return None
 
-    def get_object_for_this_type(self, **kwargs):
+    def get_object_for_this_type(self, using=None, **kwargs):
         """
         Return an object of this type for the keyword arguments given.
         Basically, this is a proxy around this object_type's get_object() model
         method. The ObjectNotExist exception, if thrown, will not be caught,
         so code that calls this method should catch it.
         """
-        return self.model_class()._base_manager.using(self._state.db).get(**kwargs)
+        if using:
+            return self.model_class()._base_manager.using(using).get(**kwargs)
```

Comment:
This path is not covered by tests. Do we need this change?

Reply:
Thanks, @felixxm. If I remove that path, I get the follow test error:
```
ERROR: test_generic_key_separation (multiple_database.tests.QueryTestCase.test_generic_key_separation)
AttributeError: 'NoneType' object has no attribute 'title'
```

Reply:
Sorry I wanted to add this comment to the second branch.

I don't see any test failures when simplifying `get_object_for_this_type()` to the:
```
def get_object_for_this_type(self, using=None, **kwargs):
    return self.model_class()._base_manager.using(using).get(**kwargs)
```

Do we need a special path when `using` is `None`? If yes, can we add test for it.

Reply:
> the base manager won't have a DB set to a different value than the router default, that might get overridden if we pass in `None` to the `using()` call

Can you add a regression test for this scenario? I think we should keep the previous behavior when `using` is not set, i.e. `self.model_class()._base_manager.using(using or self._state.db).get(**kwargs)` :thinking: What do you think?

Reply:
I tried change the code to `using(using or self._state.db)`, and I got the following error:
```
ERROR: test_contenttype_in_separate_db (multiple_database.tests.QueryTestCase.test_contenttype_in_separate_db)
multiple_database.models.Book.DoesNotExist: Book matching query does not exist.
```

I'm not really sure how to add a regression test for this scenario. Does django support having a `_base_manager` pointing to a different database than what the database router points to for those objects? Seems like some parts of django might break in that case.

But I'm not clear on all the details here. I'll look at it more, but any pointers are welcome.

Reply:
> I tried change the code to using(using or self._state.db), and I got the following error:

`book_type` in on the `default` and book is on the `other`, so `get_object_for_this_type()` doesn't return an object, so it's better to pass `None`. I'd add two more assertions for:
- `get_object_for_this_type(using="default")`
- `get_object_for_this_type(using="other")`
to test all possible options.

Reply:
I added a couple asserts... is that what you had in mind?